[1.16] Port + add features to Ender Shelf from 1.12.

trunksbomb opened the pull request:
Change to now takes actual enchanted books instead of stripping enchant from the book.
Change from 2x2 grid display to 5x1 grid -- fits an extra enchantment per block and has more room for text
Add multiblock and automation support. Insert into any single shelf and it will try to insert into other attached shelves first if an Enchanted book is present there.
Add rudimentary text scaling for long enchantment names to avoid overlap/overhang
Add Net Code for syncing from server to client on insert

Commit:
port + add features to Ender Shelf from 1.12.
Change to now takes actual enchanted books instead of stripping enchant from the book.
Change from 2x2 grid display to 5x1 grid -- fits an extra enchantment per block and has more room for text
Add multiblock and automation support. Insert into any single shelf and it will try to insert into other attached shelves first if an Enchanted book is present there.
Add rudimentary text scaling for long enchantment names to avoid overlap/overhang
Add Net Code for syncing from server to client on insert

trunksbomb commented:
Need to figure out client/server desync -- sometimes when inserting via automation the client reports more books than there really are. Logging out/in fixes, no books are actually duplicated.

Also books may disappear if all connected Ender Shelves are taken, gotta figure that out..

https://imgur.com/BOdns8U

Commit:
Fix client/server desync when inserting enchanted books into Ender Shelf
Fix shelves consuming books when there's no place for them

Commit:
Second, and I think final, fix for consuming books when there's nowhere for them to go.
Dev feature, extract from bottom of Shelf to get random encahnted books out of thin air, great to test filling up a full Ender Library.
New performance friendly function for finding nearby shelves, but doesn't know if they're connected. Needs another rewrite for performance + connected status.

Commit:
Another 2 iterations of the search for connected shelves, getting closer.

Lothrazar commented on src/main/java/com/lothrazar/cyclic/block/endershelf/EnderShelfItemHandler.java:
```diff
@@ -0,0 +1,246 @@
+      return false;
+    }).findFirst();
+    BlockPos finals = possibleOtherDestination.orElse(shelvesWithFreeSlots.stream().findFirst().orElse(originalPos));
+    //System.out.println("Going to try " + finals.toString());
```
I havent ran the code, just looking over thinking about the insert dupe when its full  

If all shelves are full and theres no spot in the book, should this return orElse(null)?  
If it returns originalPos== this.shelf.getPos() , then above  it wont hit the "else return stack" statement.

trunksbomb replied:
From what I understand, orElse isn't even supposed to take null as an argument (it's @Nonnull).

But the section of code you commented on is where we're trying to find a valid shelf to put it in, it's not actually doing any insertion here.

1. The `findShelfForInsert` method does the following:
2. Searches recursively in a defined radius for connected shelves
3. Iterate over their item handlers doing two things -- adding shelves with free slots to the `shelvesWithFreeSlots` Set and adding shelves that already have the book and room to put it to a stream.
4. At the end of that stream, `findFirst()` gets the first shelf that already has the enchantment and room for one more and stores it as an `Optional<BlockPos>`.
5. Finally, either return that single matching position of a shelf, or the first one with a free slot, or the original position if both of those fail.

Commit:
entirely refactor EnderShelfItemHandler
Add Ender Shelf Controller to act as an input/output for multiblock structure instead of the mess I was doing before
Change to compute list of attached shelves on block place/remove instead of every time item should be inserted/extracted
Add recipe, hardness, guide book entry for both
handle TE not loaded on client side when receiving packet

Commit:
handle TE not loaded on client side when receiving packet

trunksbomb commented:
Yea so I completely rewrote the port because I was doing some hacky stuff before with fake inventory slots... Here are the features now:

- Ender Shelf can be used standalone or as part of a multiblock structure with one or more Ender Shelf Controllers
- Right-click with single-enchantment enchanted book in hand on any open slot to insert the book. Stacks up to 16.
- Right-click an Ender Shelf Controller with single-enchantment enchanted book in hand to insert into the first matching slot, or first empty slot if the enchantment isn't in the multiblock library yet
- Insert enchanted books via automation -- either to a single shelf or to a Controller.

Extraction via automation not complete yet, and it still has the Dev feature of extracting random enchanted books out of thin air from the bottom-face of a Controller so it's not ready to go live, but it's definitely functional.

Commit:
Finish implementation of Ender Shelf/Controller
Add extraction from controller extracts from connected shelves

Commit:
Merge branch 'trunk/1.16' of https://github.com/Lothrazar/Cyclic into trunk/1.16_port_endershelf

 Conflicts:
	src/main/java/com/lothrazar/cyclic/registry/BlockRegistry.java
	src/main/java/com/lothrazar/cyclic/registry/ItemRegistry.java
	src/main/java/com/lothrazar/cyclic/registry/TileRegistry.java
	src/main/resources/assets/cyclic/lang/en_us.json

Commit:
refactor PacketTileInventory to conform to proxy refactor

Commit:
Comment out dev feature code

trunksbomb commented:
It's ready.

Can now extract from individual shelves or from controller via automation.
Gave the controller face a different texture just to make it easier to spot. Eventually would like to add tile renderer here to show stats about connected shelves? Maybe. It's feature complete as-is though.
